Create JavaScript article

ggliv opened the pull request:
Closes #3.

You can preview the rendered contents of this article [here](https://github.com/ggliv/DevDogs-Academy/blob/feature/js-article/frontend/javascript/README.md).

justinquinnb commented:
Already said on Discord but commenting here too just for documentation's sake:

Fantastic work so far and excited to see where this article goes. Keep up the good work and give me or Avery a ping here/on Discord as you continue developing this. We really appreciate the hard work!

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+```
+
+Undefined and null values are kind of special. Undefined is used to indicate that an identifier hasn't been given an explicit value. You'll see undefined values if you try to do something like use a variable that hasn't been assigned to. Null values, on the other hand, are used to indicate an intentional value of 'nothing'. Most of the time, you shouldn't go out of your way to use undefined. Null ...
+```js
+undefined
+null
+```
````
Do we want to cover operations on `null` and `undefined` values?

justinquinnb replied:
Unless they produce a behavior different than what someone with basic CS knowledge might expect (e.g. a value instead of an error), I think this omission is alright.

ggliv replied:
This is JavaScript, of course they do something unexpected. Jokes aside, I'll include a general statement on how operations can produce weird results and some very brief examples. I don't think it's important enough to go into much more detail than that.

ggliv commented on frontend/javascript/README.md:
```diff
@@ -0,0 +1,713 @@
+
+[^other-primitives]: Okay, technically there's also [`bigint`](https://developer.mozilla.org/en-US/docs/Glossary/BigInt) and [`symbol`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol), but most readers probably don't need to worry about those.
+
+### Variables
```
I opted to omit `var` and no-keyword declarations since i see them as largely legacy features. Should they be included?

justinquinnb replied:
As legacy features, I agree that there's little reason to include them. Let's see what @nivedhanatarajan7 thinks.

nivedhanatarajan7 replied:
Yes, adding the note about var as a footnote sounds fine. Thank you!

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+
+### Variables
+You may from time to time want to store values for later use. You can do this using variables.
+```js
+// declares a new variable `variable0` with initial value `undefined`
+let variable0;
+// variables can be reassigned
+variable0 = "hi";
+// js is dynamically typed, so variables can also be reassigned to values with
+// different types
+variable0 = 0;
+// const variables must be initialized at declaration and cannot be reassigned
+// after declaration
+// const variable1; <- not allowed!
+const variable1 = 1;
+// variable1 = 2; <- not allowed!
+
+// variables are scoped to the block they're in. this means that they're only
+// accessible inside the curly braces ({}) they were declared.
+let outerScope = "outer!";
+// outerScope now accessible
+{
+    // outerScope still accessible
+    let innerScope = "inner!";
+    // innerScope now accessible
+}
+// innerScope no longer accessible
+// outerScope still accessible
+```
````
Should we include anything on variable shadowing?

justinquinnb replied:
Can't speak to this, will leave for Avery or Nivedha.

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+// outerScope still accessible
+```
+
+Sometimes you'll want to do an operation on a variable's value and then store the result back into the variable. JavaScript has a shorthand for this:
+```js
+let foo = 100;
+// instead of writing something like this
+foo = foo + 1;
+// you can instead write this
+foo += 1;
+```
+You use this operation-then-assign syntax with any of the operations mentioned in the previous section.
````
I opted to omit mentions of the increment (`++`) and decrement (`--`) operators since I personally consider them to be a code smell. They'd also be a bit tough to describe in a concise way. Do we want to cover them?

justinquinnb replied:
This I will leave entirely for @nivedhanatarajan7 to decide on. As someone mostly experienced in backend languages like Java and C++, increment/decrement operators seem like an important inclusion, but it sounds like they may work differently in JS.

ggliv replied:
To be clear, this is very much a me thing. They behave similarly to other C langs, I just personally do not like the usage of increment and decrement. They aren't strictly necessary to do anything (just use `{+,-}=`), they only ever save like one line, and their behavior when used in ~anything outside of a for-loop is usually hard to grasp.

I'm still happy to include mention of them if others determine they need to be touched on, but IMO they don't.

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+```
+You use this operation-then-assign syntax with any of the operations mentioned in the previous section.
+
+### Comparisons
````
Should we cover null coalescing and/or optional chaining?

justinquinnb replied:
Can't speak to. Leaving open for Nivedha or Avery.

ggliv commented on frontend/javascript/README.md:
```diff
@@ -0,0 +1,713 @@
+// finally, we have the ordering comparisons. these work as you would expect in
+// any other language, with the caveat that types are always converted (like
+// with ==)
+0 > 0 // false
+0 >= "0" // true
+0 < "0" // false
+0 <= 0 // true
```
I didn't talk about the string ordering comparisons since they're not terribly common and I didn't want to type out an explanation of lex order. Should they be touched on?

justinquinnb replied:
I wouldn't want to explain lex order either! If for some reason lex order comparisons are needed by a member, maybe chuck in a link/note directing readers to an in-depth explanation of it, since it has such a niche use-case?

ggliv replied:
I think I'm comfortable letting people discover it themselves if they every need it. The JS resources listed at the bottom (or Google) should be enough for curious readers.

If anyone ever asks about it in the context of the club, we can add it pretty easily.

ggliv commented on frontend/javascript/README.md:
```diff
@@ -0,0 +1,713 @@
+nested[1][0];
+// you can get an array's length using the `.length` property (more on
+// properties in the classes section). this yields 4.
+arr.length;
```
Are there any other show-stoppingly important array methods that should be included here?

justinquinnb replied:
Will let Nivedha or Avery chime in on this.

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+true === !!"false"
+```
+
+### Conditionals
````
I omitted switch statements since they're not very common. Should they be included?

justinquinnb replied:
Since they're not commonly used in the language, I agree that their omission is justified.

ggliv commented on frontend/javascript/README.md:
```diff
@@ -0,0 +1,713 @@
+    message += "times";
+}
+
+// for-of loops are analogous to for-each loops in other languages. they
+// let you loop through collections of things.
+message = "We've seen elements ";
+let array = [0, 1, 2, 3, 4, 5];
+// trying to change the variable you bind each element to will just change
+// the variable, not the array, so we usually const-declare them for clarity
+for (const element of array) {
+    message += element + " "
+}
+// message === "We've seen elements 0 1 2 3 4 5 "
```
Should we cover destructuring with for-of?

justinquinnb replied:
Can't speak to this, will leave for Nivedha and/or Avery.

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+// message === "We've seen elements 0 1 2 3 4 5 "
+```
+
+### Functions
````
Should we talk about default parameters, parameter destructuring, varargs, generators, and/or closures?

justinquinnb replied:
I would argue for varargs and default params, but I can't speak to any of those other functionalities. Perhaps Avery or Nivedha have a stance on them.

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+lambdaSuccessor(1) === 2;
+```
+
+### Objects
````
Should we talk about getters and setters? (the JavaScript syntax sugar, not necessarily the broader concept of getters and setters)

justinquinnb replied:
If JS has a shorthand for them or a very different way of writing them than Java/other introductory languages, I would argue for including getters and setters in JS.

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+greetingObject.greetConfused("Bob") === "Goodbye Bob?";
+```
+
+### Classes
````
Should we include stuff on anonymous classes? Prototypes?

justinquinnb replied:
Will leave for Nivedha or Avery's input.

ggliv commented on frontend/javascript/README.md:
````diff
@@ -0,0 +1,713 @@
+```
+
+## Interacting with the DOM
+An important part of using JavaScript is changing and responding to changes in the DOM. The DOM, short for **D**ocument **O**bject **M**odel, is the browser's internal data structure for the contents of a page. Web browsers expose a number of JavaScript objects and functions to let you interact with this model. Understanding this section requires a basic understanding of HTML and CSS. Check out [ ...
````
We'll need to come back and make sure that these links are correct once their respective articles get written. (cc #1, #2)

justinquinnb replied:
Noted.

ggliv reviewed (commented):
I'm happy enough with the content covered in the current revision of the article, but here are some things I want to make sure are okay to omit (along with some general comments).

ggliv commented:
The article still has some rough edges that I'll come back to sand down eventually, but the content is now there and largely in the shape I want it to be in.

justinquinnb commented:
This has my utmost approval--looks great, very well done. Waiting on another set of eyes before we pull this in. Thank you for the great work!

ggliv commented:
Okay, I think I've addressed all of your review comments. I broke a few of the sections up where I could, but the conditionals section seemed mostly okay to me so I just extracted the bool ops/truthy section into its own block.

Thank you so much for taking the time to give this PR a look, I really appreciate it! Review work is far too underappreciated, I think.

nivedhanatarajan7 reviewed (approved):
The article looks great! Thank you so much!
